rowexec: refactor index join to use JoinReader

helenmhe-zz opened the pull request:
Previously there was a separate processor specifically for index joins.
This commit instantiates a joinReader processor instead, so that we
can get rid of the indexJoiner processor and indexJoinerNode.

Resolves: #40749

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/51416)
<!-- Reviewable:end -->

helenmhe commented:
Still running batch size tests with tpchvec, will update the constant when done (currently seeing a big performance hit in the multinode case)

asubiotto commented:
Nice! Haven't had the chance to take a look yet. When you push up a WIP, you can still add `Reviewers` to your PR (or at least cc @) even though you're just looking for early feedback so that the team is notified via email about the PR.

yuzefovich reviewed (commented):
Nice to see this work coming along!

I have some questions and some nits. Also, shouldn't `indexjoiner.go` file be removed?

Also I see that there are some formatting failures, are you using the `crlfmt` in Goland? Here is the link to how I set it up, https://cockroachlabs.slack.com/archives/C4VM697PC/p1545384791000100?thread_ts=1540908498.001700&cid=C4VM697PC.

nit: in the commit title, `s/sql/rowexec/` since this touches mostly rowexec package.

> Reviewed 5 of 5 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51416#-:-MCJ86Hi6T7HXVcVlTeG:b-j147bm)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto and @helenmhe)

___
*[pkg/sql/rowexec/joinreader.go, line 102 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51416#-MCJ8KsJ2Lt35hoXnkti:-MCJ8KsJ2Lt35hoXnktj:b-1oxclx) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf7a6e6cbd8e893e5501a49eed5b7c34d0d4136a/pkg/sql/rowexec/joinreader.go#L102)):*
> ```Go
> 	post *execinfrapb.PostProcessSpec,
> 	output execinfra.RowReceiver,
> 	indexJoin bool,
> ```

Our style guide discourages the usage of `bool` arguments, and I think in this case it is warranted to define a special enum, something like
```
type joinReaderType int

const (
  lookupJoinReaderType joinReaderType = iota
  indexJoinReaderType
)
```
and use this enum in place of boolean.
___
*[pkg/sql/rowexec/joinreader.go, line 105 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51416#-MCJ8lFUEiKoM4So7vHV:-MCJ8lFUEiKoM4So7vHW:bwk4yp8) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf7a6e6cbd8e893e5501a49eed5b7c34d0d4136a/pkg/sql/rowexec/joinreader.go#L105)):*
> ```Go
> ) (execinfra.RowSourcedProcessor, error) {
> 	if spec.IndexIdx != 0 && indexJoin {
> 		return nil, errors.Errorf("index join must be against primary index")
> ```

nit: I'd use `errors.AssertionFailed` because such case is really unexpected (it will include a stack trace which could help with the debugging).
___
*[pkg/sql/rowexec/joinreader.go, line 152 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51416#-MCJ9Kx03KOy-xSPRaiM:-MCJ9Kx03KOy-xSPRaiN:b99kd6i) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf7a6e6cbd8e893e5501a49eed5b7c34d0d4136a/pkg/sql/rowexec/joinreader.go#L152)):*
> ```Go
> 
> 	if indexJoin {
> 		if err := jr.joinerBase.init(
> ```

I think this could refactored so that we have only a single `jr.joinerBase.init` call. Probably we'll need to define some local variables.
___
*[pkg/sql/rowexec/joinreader.go, line 213 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51416#-MCJ9aJ_-zSv89wB_3q0:-MCJ9aJ_-zSv89wB_3q1:b5t03u4) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf7a6e6cbd8e893e5501a49eed5b7c34d0d4136a/pkg/sql/rowexec/joinreader.go#L213)):*
> ```Go
> 	var fetcher row.Fetcher
> 	if indexJoin {
> 		_, _, err = initRowFetcher(
> ```

ditto for refactoring - I think we can have a single `initRowFetcher` call.
___
*[pkg/sql/rowexec/joinreader.go, line 251 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51416#-MCJ9lwo5Qrnv0HEfzgq:-MCJ9lwo5Qrnv0HEfzgr:b4fa2cm) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf7a6e6cbd8e893e5501a49eed5b7c34d0d4136a/pkg/sql/rowexec/joinreader.go#L251)):*
> ```Go
> 	spanBuilder := span.MakeBuilder(flowCtx.Codec(), &jr.desc, jr.index)
> 	if indexJoin {
> 		spanBuilder.SetNeededColumns(jr.Out.NeededColumns())
> ```

nit: it might make sense to taking in `neededColumns` as an argument in `initJoinReaderStrategy` method.
___
*[pkg/sql/rowexec/joinreader.go, line 439 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51416#-MCJ9ygD-OK-G8jPqq9z:-MCJ9ygD-OK-G8jPqqA-:b4fpn2w) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf7a6e6cbd8e893e5501a49eed5b7c34d0d4136a/pkg/sql/rowexec/joinreader.go#L439)):*
> ```Go
> 	// restore the original order of the output during the output collection
> 	// phase.
> 	if jr.shouldLimitBatches {
> ```

This looks suspicious, maybe a better check is whether we're doing a lookup join? And what happens when we do sort the spans with an index join?
___
*[pkg/sql/rowexec/joinreader_strategies.go, line 271 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51416#-MCJBQkABRtk9-Z2TkQ-:-MCJBQkABRtk9-Z2TkQ0:b-s9hmw) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf7a6e6cbd8e893e5501a49eed5b7c34d0d4136a/pkg/sql/rowexec/joinreader_strategies.go#L271)):*
> ```Go
> // offered a significant performance improvement. Larger batch sizes offered
> // small to no marginal improvements.
> func (s *joinReaderNoOrderingStrategy) getLookupRowsBatchSizeHint() int64 {
> ```

super nit: it might nice to keep all method implementations in the same order as they are defined in `joinReaderStrategy` interface.
___
*[pkg/sql/rowexec/joinreader_strategies.go, line 327 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51416#-MCJAH-sEeHSFBWqr5Zi:-MCJAH-sEeHSFBWqr5Zj:b2b75i8) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf7a6e6cbd8e893e5501a49eed5b7c34d0d4136a/pkg/sql/rowexec/joinreader_strategies.go#L327)):*
> ```Go
> func (s *joinReaderIndexJoinStrategy) close(ctx context.Context) {}
> 
> func (s *joinReaderIndexJoinStrategy) getLookupRowsBatchSizeHint() int64 {
> ```

nit: this deserves a comment on how this number was obtained.
___
*[pkg/sql/rowexec/joinreader_strategies.go, line 328 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51416#-MCJAh9vCiaSpZEM1q9U:-MCJAh9vCiaSpZEM1q9V:b-1fs0q5) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf7a6e6cbd8e893e5501a49eed5b7c34d0d4136a/pkg/sql/rowexec/joinreader_strategies.go#L328)):*
> ```Go
> 
> func (s *joinReaderIndexJoinStrategy) getLookupRowsBatchSizeHint() int64 {
> 	return 2 << 21
> ```

super nit: it might be slightly more clean to use `4 << 20 /* 4MB */` here.


<!-- Sent from Reviewable.io -->

helenmhe-zz reviewed (commented):
Yeah I'm using `crlfmt` so not sure if it's broken - when I tried running the linter locally (`make lint TESTS='TestGolint'`) it said everything was fine, is there something else I can run?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51416#-:-MCJCgSLB8vhyHiR7m2p:bmsw8jh)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto and @helenmhe)



<!-- Sent from Reviewable.io -->

yuzefovich commented:
Sometimes it seems like the file watcher is breaking (maybe when Goland is updated?), so in those cases I remove `crlfmt` and add back from the scratch, and then everything works.

helenmhe-zz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51416)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto, @helenmhe, and @yuzefovich)

___
*[pkg/sql/rowexec/joinreader.go, line 439 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51416#-MCJ9ygD-OK-G8jPqq9z:-MCJFHHg-IO3MVPYv7jp:b-1w24yk) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf7a6e6cbd8e893e5501a49eed5b7c34d0d4136a/pkg/sql/rowexec/joinreader.go#L439)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

This looks suspicious, maybe a better check is whether we're doing a lookup join? And what happens when we do sort the spans with an index join?
</blockquote></details>

When the spans are sorted in an index join, the output rows are emitted in the wrong order for some tests


<!-- Sent from Reviewable.io -->

helenmhe-zz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51416)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto and @yuzefovich)

___
*[pkg/sql/rowexec/joinreader.go, line 102 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51416#-MCJ8KsJ2Lt35hoXnkti:-MCJKGtmCcP8-quNtmlB:b-t9490a) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf7a6e6cbd8e893e5501a49eed5b7c34d0d4136a/pkg/sql/rowexec/joinreader.go#L102)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Our style guide discourages the usage of `bool` arguments, and I think in this case it is warranted to define a special enum, something like
```
type joinReaderType int

const (
  lookupJoinReaderType joinReaderType = iota
  indexJoinReaderType
)
```
and use this enum in place of boolean.
</blockquote></details>

added!
___
*[pkg/sql/rowexec/joinreader.go, line 105 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51416#-MCJ8lFUEiKoM4So7vHV:-MCJJtfu80pHj4aXD4Dk:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf7a6e6cbd8e893e5501a49eed5b7c34d0d4136a/pkg/sql/rowexec/joinreader.go#L105)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

nit: I'd use `errors.AssertionFailed` because such case is really unexpected (it will include a stack trace which could help with the debugging).
</blockquote></details>

done
___
*[pkg/sql/rowexec/joinreader.go, line 152 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51416#-MCJ9Kx03KOy-xSPRaiM:-MCJJvRK0lA-zPxsTonj:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf7a6e6cbd8e893e5501a49eed5b7c34d0d4136a/pkg/sql/rowexec/joinreader.go#L152)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

I think this could refactored so that we have only a single `jr.joinerBase.init` call. Probably we'll need to define some local variables.
</blockquote></details>

done
___
*[pkg/sql/rowexec/joinreader.go, line 213 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51416#-MCJ9aJ_-zSv89wB_3q0:-MCJKe8gDvVxHKBEJ58h:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf7a6e6cbd8e893e5501a49eed5b7c34d0d4136a/pkg/sql/rowexec/joinreader.go#L213)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

ditto for refactoring - I think we can have a single `initRowFetcher` call.
</blockquote></details>

done
___
*[pkg/sql/rowexec/joinreader.go, line 251 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51416#-MCJ9lwo5Qrnv0HEfzgq:-MCJKagL0xE-0Cnd_fy3:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf7a6e6cbd8e893e5501a49eed5b7c34d0d4136a/pkg/sql/rowexec/joinreader.go#L251)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

nit: it might make sense to taking in `neededColumns` as an argument in `initJoinReaderStrategy` method.
</blockquote></details>

done
___
*[pkg/sql/rowexec/joinreader_strategies.go, line 271 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51416#-MCJBQkABRtk9-Z2TkQ-:-MCJK28U9CFISBw2Ew5b:b903zt9) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf7a6e6cbd8e893e5501a49eed5b7c34d0d4136a/pkg/sql/rowexec/joinreader_strategies.go#L271)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

super nit: it might nice to keep all method implementations in the same order as they are defined in `joinReaderStrategy` interface.
</blockquote></details>

whoops fixed
___
*[pkg/sql/rowexec/joinreader_strategies.go, line 327 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51416#-MCJAH-sEeHSFBWqr5Zi:-MCJKnJP2Ul2b2v83wT_:b-7tfmjp) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf7a6e6cbd8e893e5501a49eed5b7c34d0d4136a/pkg/sql/rowexec/joinreader_strategies.go#L327)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

nit: this deserves a comment on how this number was obtained.
</blockquote></details>

added
___
*[pkg/sql/rowexec/joinreader_strategies.go, line 328 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51416#-MCJAh9vCiaSpZEM1q9U:-MCJKlme1Lv8EvE20Pb7:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf7a6e6cbd8e893e5501a49eed5b7c34d0d4136a/pkg/sql/rowexec/joinreader_strategies.go#L328)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

super nit: it might be slightly more clean to use `4 << 20 /* 4MB */` here.
</blockquote></details>

done


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
nit: the convention `s/sql/rowexec/` means "replacing `sql` with `rowexec`" (https://www.geeksforgeeks.org/sed-command-in-linux-unix-with-examples/).

Also, I think `TestIndexJoiner` from `indexjoiner_test.go` should be moved into `joinreader_test.go`.

> Reviewed 5 of 5 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51416#-:-MCJUngz-BxLvRRjVexM:b-jfz0ci)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto and @helenmhe)

___
*[pkg/sql/rowexec/joinreader.go, line 439 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51416#-MCJ9ygD-OK-G8jPqq9z:-MCJWREP6nPpn1v7Ct8z:bwbw0w7) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf7a6e6cbd8e893e5501a49eed5b7c34d0d4136a/pkg/sql/rowexec/joinreader.go#L439)):*
<details><summary><i>Previously, helenmhe (Helen He) wrote…</i></summary><blockquote>

When the spans are sorted in an index join, the output rows are emitted in the wrong order for some tests
</blockquote></details>

Hm, what's an example? Maybe the queries don't have `ORDER BY` clause so the output row order is actually non-deterministic? If that's the case, in logic tests we have `rowsort` directive to ignore the output row order (alternatively, we could add `ORDER BY` to the queries).

AFAIU it is not required for the correctness of the index join to not change the order of spans here, however, at the same time that sort is not needed in the index join.
___
*[pkg/sql/rowexec/joinreader.go, line 117 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/51416#-MCJVVaHCQW1qQilzS0l:-MCJVVaIBHAaJ-pKDi6r:bfxu0wy) ([raw file](https://github.com/cockroachdb/cockroach/blob/bec58d5c0d8b0bbf7c085535d2b9308c33b955e9/pkg/sql/rowexec/joinreader.go#L117)):*
> ```Go
> 
> 	var jr *joinReader
> 	if readerType == lookupJoinReaderType {
> ```

nit: I think the conventional way to handle enums is to use a switch:
```
switch readerType {
  case lookupJoinReaderType:
  ...
  case indexJoinReaderType:
  ...
  default:
    panic("unsupported joinReaderType")
}
```
However, at the same time I don't think that we'll be adding any other values to `joinReaderType` enums, so maybe it's reasonable to use `if` statements as you currently do. Up to you.
___
*[pkg/sql/rowexec/joinreader.go, line 130 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/51416#-MCJVr3L-U6Otphmn_M-:-MCJVr3L-U6Otphmn_M0:b-ehqh2z) ([raw file](https://github.com/cockroachdb/cockroach/blob/bec58d5c0d8b0bbf7c085535d2b9308c33b955e9/pkg/sql/rowexec/joinreader.go#L130)):*
> ```Go
> 			pkCols[i] = uint32(i)
> 		}
> 		jr = &joinReader{
> ```

nit: I think we could also refactor this to have a single `joinReader` object instantiation since only `lookupCols` parameter differs.
___
*[pkg/sql/rowexec/joinreader.go, line 227 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/51416#-MCJWC914DVYPGRa7JDk:-MCJWC914DVYPGRa7JDl:b-spzy4b) ([raw file](https://github.com/cockroachdb/cockroach/blob/bec58d5c0d8b0bbf7c085535d2b9308c33b955e9/pkg/sql/rowexec/joinreader.go#L227)):*
> ```Go
> 
> 	jr.initJoinReaderStrategy(
> 		flowCtx, jr.desc.ColumnTypesWithMutations(returnMutations), len(columnIDs), spec.MaintainOrdering, neededRightCols,
> ```

super nit: seems unusual to break the big line in such a fashion - we strive for lines with up to 80 characters, so I'd insert a newline sooner.


<!-- Sent from Reviewable.io -->

helenmhe-zz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51416)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto, @helenmhe, and @yuzefovich)

___
*[pkg/sql/rowexec/joinreader.go, line 439 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51416#-MCJ9ygD-OK-G8jPqq9z:-MCMxrrVAEnSXWSZiU0h:b-bqxbvx) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf7a6e6cbd8e893e5501a49eed5b7c34d0d4136a/pkg/sql/rowexec/joinreader.go#L439)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Hm, what's an example? Maybe the queries don't have `ORDER BY` clause so the output row order is actually non-deterministic? If that's the case, in logic tests we have `rowsort` directive to ignore the output row order (alternatively, we could add `ORDER BY` to the queries).

AFAIU it is not required for the correctness of the index join to not change the order of spans here, however, at the same time that sort is not needed in the index join.
</blockquote></details>

Some of the schema change tests were failing due to row order like this:

```  
    TestSchemaChangeAfterCreateInTxn: schema_changer_test.go:4428: e = 0, c = 137, v = 133
    TestSchemaChangeAfterCreateInTxn: schema_changer_test.go:4428: e = 1, c = 136, v = 132
    TestSchemaChangeAfterCreateInTxn: schema_changer_test.go:4428: e = 2, c = 135, v = 131
...
    TestSchemaChangeAfterCreateInTxn: schema_changer_test.go:4428: e = 131, c = 6, v = 2
    TestSchemaChangeAfterCreateInTxn: schema_changer_test.go:4428: e = 132, c = 5, v = 1
    TestSchemaChangeAfterCreateInTxn: schema_changer_test.go:4428: e = 133, c = 4, v = 0
    TestSchemaChangeAfterCreateInTxn: schema_changer_test.go:4428: e = 134, c = 271, v = 267
    TestSchemaChangeAfterCreateInTxn: schema_changer_test.go:4428: e = 135, c = 270, v = 266
    TestSchemaChangeAfterCreateInTxn: schema_changer_test.go:4428: e = 136, c = 269, v = 265```

I just checked and the query doesn't have an `ORDER BY` so seems like that was the issue?
```



<!-- Sent from Reviewable.io -->

helenmhe-zz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51416)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto and @yuzefovich)

___
*[pkg/sql/rowexec/joinreader.go, line 117 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/51416#-MCJVVaHCQW1qQilzS0l:-MCN4eQxDIoDInDVe3uF:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/bec58d5c0d8b0bbf7c085535d2b9308c33b955e9/pkg/sql/rowexec/joinreader.go#L117)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

nit: I think the conventional way to handle enums is to use a switch:
```
switch readerType {
  case lookupJoinReaderType:
  ...
  case indexJoinReaderType:
  ...
  default:
    panic("unsupported joinReaderType")
}
```
However, at the same time I don't think that we'll be adding any other values to `joinReaderType` enums, so maybe it's reasonable to use `if` statements as you currently do. Up to you.
</blockquote></details>

done
___
*[pkg/sql/rowexec/joinreader.go, line 130 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/51416#-MCJVr3L-U6Otphmn_M-:-MCN4ZRe8hdYi03beF5j:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/bec58d5c0d8b0bbf7c085535d2b9308c33b955e9/pkg/sql/rowexec/joinreader.go#L130)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

nit: I think we could also refactor this to have a single `joinReader` object instantiation since only `lookupCols` parameter differs.
</blockquote></details>

done
___
*[pkg/sql/rowexec/joinreader.go, line 227 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/51416#-MCJWC914DVYPGRa7JDk:-MCMzrc64XDDDkLxaFsU:b-6hc81r) ([raw file](https://github.com/cockroachdb/cockroach/blob/bec58d5c0d8b0bbf7c085535d2b9308c33b955e9/pkg/sql/rowexec/joinreader.go#L227)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

super nit: seems unusual to break the big line in such a fashion - we strive for lines with up to 80 characters, so I'd insert a newline sooner.
</blockquote></details>

whoops my goland was set to 120, fixed


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 4 of 4 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51416#-:-MCNMeId0zdDgX1ee5T4:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @asubiotto and @helenmhe)

___
*[pkg/sql/rowexec/joinreader.go, line 439 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51416#-MCJ9ygD-OK-G8jPqq9z:-MCNKw8I603nInpBy3Mw:bfb4owt) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf7a6e6cbd8e893e5501a49eed5b7c34d0d4136a/pkg/sql/rowexec/joinreader.go#L439)):*
<details><summary><i>Previously, helenmhe (Helen He) wrote…</i></summary><blockquote>

Some of the schema change tests were failing due to row order like this:

```  
    TestSchemaChangeAfterCreateInTxn: schema_changer_test.go:4428: e = 0, c = 137, v = 133
    TestSchemaChangeAfterCreateInTxn: schema_changer_test.go:4428: e = 1, c = 136, v = 132
    TestSchemaChangeAfterCreateInTxn: schema_changer_test.go:4428: e = 2, c = 135, v = 131
...
    TestSchemaChangeAfterCreateInTxn: schema_changer_test.go:4428: e = 131, c = 6, v = 2
    TestSchemaChangeAfterCreateInTxn: schema_changer_test.go:4428: e = 132, c = 5, v = 1
    TestSchemaChangeAfterCreateInTxn: schema_changer_test.go:4428: e = 133, c = 4, v = 0
    TestSchemaChangeAfterCreateInTxn: schema_changer_test.go:4428: e = 134, c = 271, v = 267
    TestSchemaChangeAfterCreateInTxn: schema_changer_test.go:4428: e = 135, c = 270, v = 266
    TestSchemaChangeAfterCreateInTxn: schema_changer_test.go:4428: e = 136, c = 269, v = 265```

I just checked and the query doesn't have an `ORDER BY` so seems like that was the issue?
</blockquote></details>

Yeah, probably.

But I also think it's ok to have this `if` check, I'd only move the comment right above inside of the `if` block.
___
*[pkg/sql/rowexec/joinreader.go, line 116 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/51416#-MCNMRVvCnF7XCZb3yPm:-MCNMRVvCnF7XCZb3yPn:b-d27dc5) ([raw file](https://github.com/cockroachdb/cockroach/blob/800a1ab91fca6351be25bb42fdeea3b56879f793/pkg/sql/rowexec/joinreader.go#L116)):*
> ```Go
> 	}
> 
> 	var jr *joinReader
> ```

nit: no need to define `jr` here.
___
*[pkg/sql/rowexec/joinreader.go, line 155 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/51416#-MCNLpNfAPzZWj9yh9l0:-MCNLpNfAPzZWj9yh9l1:b-g8t9mk) ([raw file](https://github.com/cockroachdb/cockroach/blob/800a1ab91fca6351be25bb42fdeea3b56879f793/pkg/sql/rowexec/joinreader.go#L155)):*
> ```Go
> 	// If the lookup columns form a key, there is only one result per lookup, so the fetcher
> 	// should parallelize the key lookups it performs.
> 	jr.shouldLimitBatches = !spec.LookupColumnsAreKey && (readerType != indexJoinReaderType)
> ```

super nit: I'd use `readerType == lookupJoinReaderType`.


<!-- Sent from Reviewable.io -->

helenmhe commented:
TFTR!

bors r+

rohany commented on pkg/sql/rowexec/joinreader.go:
```diff
@@ -99,12 +107,30 @@ func newJoinReader(
+	case lookupJoinReaderType:
+		lookupCols = spec.LookupColumns
+	default:
+		panic("unsupported joinReaderType")
```
we shouldn't panic here -- it should return error

rohany commented on pkg/sql/rowexec/joinreader.go:
```diff
@@ -162,10 +201,23 @@ func newJoinReader(
 	var fetcher row.Fetcher
+	var colIdxMap map[sqlbase.ColumnID]int
+	var rightCols util.FastIntSet
+	switch readerType {
```
Is it possible to not have to do this branch many different spots? It's fine if not.

rohany reviewed (commented):
flying by with some nits, had to study this PR for a conflicting change.

helenmhe-zz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51416)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @asubiotto, @rohany, and @yuzefovich)

___
*[pkg/sql/rowexec/joinreader.go, line 439 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51416#-MCJ9ygD-OK-G8jPqq9z:-MCYb-4uDsxLe9MUq4G_:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf7a6e6cbd8e893e5501a49eed5b7c34d0d4136a/pkg/sql/rowexec/joinreader.go#L439)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Yeah, probably.

But I also think it's ok to have this `if` check, I'd only move the comment right above inside of the `if` block.
</blockquote></details>

Done.
___
*[pkg/sql/rowexec/joinreader.go, line 155 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/51416#-MCNLpNfAPzZWj9yh9l0:-MCYarB630XCQtYCqXzF:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/800a1ab91fca6351be25bb42fdeea3b56879f793/pkg/sql/rowexec/joinreader.go#L155)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

super nit: I'd use `readerType == lookupJoinReaderType`.
</blockquote></details>

Done.
___
*[pkg/sql/rowexec/joinreader.go, line 50 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/51416#-MCCwRiFVLwVgWGO4CWW-r4-50:-MCgkWNDEjuuWQnQduje:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5ef4872076e0b4640492fd02019eb88b3e112729/pkg/sql/rowexec/joinreader.go#L50)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

[nit]: The enum type and its elements should have short comments
</blockquote></details>

Done.
___
*[pkg/sql/rowexec/joinreader.go, line 127 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/51416#-MCCwRiFVLwVgWGO4CWW-r4-127:-MCgkYf_-y7V5a9W2KWQ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5ef4872076e0b4640492fd02019eb88b3e112729/pkg/sql/rowexec/joinreader.go#L127)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

we shouldn't panic here -- it should return error
</blockquote></details>

Done.
___
*[pkg/sql/rowexec/joinreader.go, line 206 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/51416#-MCCwRiFVLwVgWGO4CWW-r4-206:-MCglMCV3yNCUtAhU9-a:b-nv1bec) ([raw file](https://github.com/cockroachdb/cockroach/blob/5ef4872076e0b4640492fd02019eb88b3e112729/pkg/sql/rowexec/joinreader.go#L206)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

Is it possible to not have to do this branch many different spots? It's fine if not.
</blockquote></details>

Unfortunately not unless I have a lot more duplicated within the cases


<!-- Sent from Reviewable.io -->

helenmhe committed:
rowexec: refactor index join to use JoinReader

Previously there was a separate processor specifically for index joins.
This commit instantiates a joinReader processor instead, so that we
can get rid of the indexJoiner processor and indexJoinerNode.

Resolves: #40749

Release note: None

yuzefovich reviewed (commented):
> Reviewed 1 of 2 files at r4, 3 of 3 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51416)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @asubiotto, @helenmhe, and @rohany)

___
*[pkg/sql/logictest/testdata/logic_test/tpch_vec, line 621 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/51416#-MCnNkWk2iHhzB7WCMjJ:-MCnNkWk2iHhzB7WCMjK:b-ynq227) ([raw file](https://github.com/cockroachdb/cockroach/blob/a9eace845854ea9d8603abeb48b03c03ead9c73e/pkg/sql/logictest/testdata/logic_test/tpch_vec#L621)):*
> ```
>             │ │ └ *rowexec.joinReader
>             │ │   └ *colexec.hashJoiner
>             │ │     ├ *colfetcher.colBatchScan
> ```

This will fail, you should run `make testbaselogic FILES=tpch_vec TESTFLAGS='-rewrite'` after rebasing.


<!-- Sent from Reviewable.io -->

helenmhe-zz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51416)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @asubiotto, @rohany, and @yuzefovich)

___
*[pkg/sql/logictest/testdata/logic_test/tpch_vec, line 621 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/51416#-MCnNkWk2iHhzB7WCMjJ:-MCnOXtU6XoQhvSff9HZ:bykgvsx) ([raw file](https://github.com/cockroachdb/cockroach/blob/a9eace845854ea9d8603abeb48b03c03ead9c73e/pkg/sql/logictest/testdata/logic_test/tpch_vec#L621)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

This will fail, you should run `make testbaselogic FILES=tpch_vec TESTFLAGS='-rewrite'` after rebasing.
</blockquote></details>

ah right thanks


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
FYI there is another PR #51636 that is currently being merged by bors that will have a merge conflict with this PR on `tpch_vec` file.

> Reviewed 2 of 2 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51416#-:-MCr7Yyn8t6BBvG6sHBa:bogg0kg)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @asubiotto and @rohany)



<!-- Sent from Reviewable.io -->

helenmhe commented:
Should I wait for the other PR to get merged through bors or try to fix the conflict ahead of time? It looks like bors is taking a while





<!-- Sent from Reviewable.io -->

yuzefovich commented:
You should wait because there is a chance that your PR gets merged first which will break things.

helenmhe commented:
TFTRs!

bors r+
